### DISCUSSION TITLE
MOGA no. of points in the populations<N>.dat files doesn't match size of evaluation blocks

### DISCUSSION THREAD

**Discussion by simonbance** [snl-dakota/dakota#51] [Q&A]
I am trying understand what is going on with the output files from my (JEGA) MOGA routine. 

There is an initial population that is decided randomly. After the objectives are returned, each point is assessed based on the "fitness" criteria and a certain number are carried forward to the next population. The new population also gets some new points, which are based on the old ones and some genetic mutations. 

This much is clear. However, I am confused as to how these populations correspond to the Blocks of evaluation points that Dakota sends to my blackbox script. The Block sizes at each generation do not match the number of new points in the corresponding population. 

Output from my python script "plot_pops.py":
```
Population files after sorting:
 ['./population_0.dat', './population_1.dat', './population_2.dat', './population_3.dat', './population_4.dat']

Loading population files...
Total 1162 data points in the population files.

There are 1162 total data points in all population files. 
618 of those points are unique! 
pop_1: 32 points (0 new).	32 points (32 unique) from previous populations. 
pop_2: 118 points (86 new).	64 points (32 unique) from previous populations. 
pop_3: 395 points (288 new).	182 points (118 unique) from previous populations. 
pop_4: 585 points (212 new).	577 points (406 unique) from previous populations. 
total_running_total: 1162
unique_running_total: 618

There are 666 data points in the moga.dat file. 
666 of those points are unique! 
```
Also, looking at output.log:
``` 
$ grep Blocking output.log
Blocking synchronize of 32 asynchronous WINBAT evaluations
Blocking synchronize of 86 asynchronous WINBAT evaluations
Blocking synchronize of 302 asynchronous WINBAT evaluations
Blocking synchronize of 246 asynchronous WINBAT evaluations
```
As you can see, the number of new (unique) points in each population do not match the number of points in each block. 

At the end of the routine, the total number of points evaluated within these blocks correctly corresponds to my specific maximum number of evaluations. That's good. But what is it that is being stored in the population_<N>.dat files, and why do these not contain the same number of *unique* points as the evaluation block sizes? 

This behaviour is even more interesting if you reduce the shrinkage factor (e.g. to 0.65) since then the total number of unique points in all the populations is smaller than the total number of evaluations. 

I'm attaching all my files for reference and would appreciate some guidance. Thank you! 
[moga_blackbox_python_dummy.zip](https://github.com/snl-dakota/dakota/files/12231056/moga_blackbox_python_dummy.zip)

**Comment by jadamstephens** [comment]
Simon,

Please look at the Theory section of the [moga keyword documenation](https://snl-dakota.github.io/docs/latest_release/users/usingdakota/reference/method-soga.html). Roughly speaking, in each generation, candidate population members for the next generation are created by crossover and possibly by mutation. These are evaluated using your simulation. Based on their fitness, they may be introduced into the population for the next generation, but they may also be discarded.

The fact that they may be discarded may be the answer to your question. Dakota evaluates candidates that don't make the cut. So, there are fewer unique members of the populations than there were Dakota evaluations.

Is that helpful, or am I missing your point?
